Instance Segmentation Confusion Matrix

mayankagarwals opened the pull request:
# Description

Solves for https://github.com/roboflow/supervision/issues/238

## Type of change

Please delete options that are not relevant.

-   [ ] Bug fix (non-breaking change which fixes an issue)
-   [x] New feature (non-breaking change which adds functionality)
-   [ ] This change requires a documentation update

## How has this change been tested, please provide a testcase or example of how you tested the change?

Please use this colab for testing: https://colab.research.google.com/drive/1OKjyH8JmFMROJ97447DUn13p-GDP6dXl
Test cases have been added

SkalskiP commented on supervision/detection/utils.py:
```diff
@@ -48,6 +48,25 @@ def box_area(box):
     return area_inter / (area_true[:, None] + area_detection - area_inter)
 
 
+def mask_iou_batch(masks_true: np.ndarray, masks_detection: np.ndarray) -> np.ndarray:
```
@mayankagarwals let's add unit tests

mayankagarwals replied:
Hey! Yep, waiting for some discussion to happen with respect to API here https://github.com/roboflow/supervision/issues/238 before completing the feature

SkalskiP commented:
Hi @mayankagarwals 👋🏻! Is that code ready for review? It is still marked as `draft`.

mayankagarwals commented:
Hey @SkalskiP. No, the code is still not complete as I'm waiting for your input on https://github.com/roboflow/supervision/issues/238. There are two different ways we can design the API.

SkalskiP commented:
@mayankagarwals okey I'm rolling that PR over to `supervision-0.14.0`

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/roboflow/supervision?pullRequest=240) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/roboflow/supervision?pullRequest=240) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/roboflow/supervision?pullRequest=240) it.</sub>

mayankagarwals commented:
Hi @kirilllzaitsev @hardikdava  

While the code here is functional, I wanted to ask your opinion on `force_masks`

Right now I'm assuming if force_masks is passed we will enforce that every detection in the `from_detections` is accompanied by a mask. 
The segmentation confusion matrix is computed only when force_masks is passed

kirilllzaitsev commented:
@mayankagarwals hey, [this PR](https://github.com/roboflow/supervision/pull/265) should be relevant. `force_masks=True ` refers to the case when the masks are enforced, while if `force_masks=False`, but the masks are present in annotations ('segmentation' field), they will be parsed since now

mayankagarwals commented:
@SkalskiP @hardikdava 
PR is ready for review. Have added a colab link for testing.

Borda committed:
feat(metrics): support MetricTarget.MASKS in ConfusionMatrix

Adds segmentation confusion-matrix support by mirroring the existing
ORIENTED_BOUNDING_BOXES dispatch pattern: masks are threaded as a parallel
optional argument (they can't flatten into the tensor row like OBB
coordinates), and evaluate_detection_batch/_split_detections_by_outcome
dispatch to mask_iou_batch. Reuses the existing vectorized mask_iou_batch
(matmul-based, chunked, zero-guarded) rather than introducing a new one.

Removes _assert_supported_target, which previously rejected MASKS
unconditionally at every entry point. Converts the 3 rejection tests to
assert ValueError only when masks are actually missing, and adds
positive coverage for perfect overlap, mask-IoU-vs-box-IoU divergence,
and IoU-threshold sensitivity.

Supersedes PR #240's force_masks/segmentationMatrix approach, which
targeted the pre-src/ layout and predates this metric_target axis.

kemo04 commented:
## AI Code Review (PR Review Agent)

- **[Logic & Correctness · High]** — `supervision/metrics/detection.py` (diff line 53): Removing the f-string prefix while keeping f-string interpolation syntax results in a literal string containing unparsed braces instead of the intended values.
